Fix LockFreeReaderHashtable

davidwrighton opened the pull request:
- Communication of the presence of an expanding hashtable was very close to not safe
  - Changed to aggressively use Interlocked operations instead
- If there were 3 threads, 1 which was expanding (thread A), 1 which failed an insert and was waiting on an expand(thread B), and a third which did a lookup for the item in process of insert on thread B, the table would find the item in process of insertion
  - Fixed by moving to a scheme where insertion didn't actually write in the value until it was confirmed that that spot was the right place to insert, and the value would be preserved into an expansion
  - In order to acquire a sentinel for scenarios without requiring the explicit definition of one, the first item inserted is used as the sentinel. It simply exists outside of the hashtable in a side variable

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -169,38 +185,83 @@ public bool TryGetValue(TKey key, out TValue value)
+        {
+            TValue sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+            if (sentinel == null)
+                return null;
```
seems like this cannot happen. we must have seen sentinel, if we are here. perhaps an assert instead

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -299,6 +376,13 @@ private TValue AddOrGetExistingInner(TValue value, out bool addedValue)
                 throw new ArgumentNullException();
 
+            if (Interlocked.CompareExchange(ref _entryInProcessOfWritingSentinel, value, null) == null)
+            {
```
It may be worth to check if _entryInProcessOfWritingSentinel is already not null before doing CAS.

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             // Now that we've written to the local array, find out if that array has been
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
+            if (Interlocked.CompareExchange(ref _newHashTable, hashTableLocal, hashTableLocal) != hashTableLocal)
```
>Interlocked.CompareExchange [](start = 16, length = 27)

this could be ordinary read/compare.  `TryWriteSentinelToLocation` used full fence to write, so the value you read here will be newer than that write. 
If expansion started before that write, you will see the new table. 
Both writes are full fences, so there is a total order. 

I.E. - it is important that sentinel and new table publishing are full fences, but this read could be a regular read.

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
+            // Add to hash, use a volatile write to ensure that
+            // the contents of the value are fully published to all
+            // threads before adding to the hashtable
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], null, _entryInProcessOfWritingSentinel) == _entryInProcessOfWritingSentinel)
```
>Interlocked.CompareExchange [](start = 16, length = 27)

This could be an ordinary write. Unless you specifically do this for the assert in the caller.

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             // the contents of the value are fully published to all
             // threads before adding to the hashtable
-            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, null) == null)
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, _entryInProcessOfWritingSentinel) == _entryInProcessOfWritingSentinel)
```
>Interlocked.CompareExchange [](start = 16, length = 27)

Same here as in `TryWriteAbortNullToLocation`. Could be an ordinary write. If want to assert for the original value, could just read it before writing. You've "claimed" the slot by writing sentinel to it so should be the only writer of the actual value or null.

And if the exclusivity is violated, ordinary read/write will catch that too.

VSadov commented:
>             }

You may want to read _hashtable before _count, or you may resize unnecessarily. 

---
Refers to: src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:392 in 89178c1. [](commit_id = 89178c1803d6dc797cc6ca480459bc758db4eea0, deletion_comment = False)

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -517,13 +696,16 @@ public Enumerator GetEnumerator()
 
             public bool MoveNext()
             {
+                if (_sentinel != null)
```
>if (_sentinel != null) [](start = 16, length = 22)

strange syntax. adding   {  }   would make it easier to read.

VSadov commented:
>                             _current = _hashtableContentsToEnumerate[_index];

Not sure if there are extra guarantees of not resizing while iterating. Otherwise we may yield transient sentinels.

---
Refers to: src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:715 in 89178c1. [](commit_id = 89178c1803d6dc797cc6ca480459bc758db4eea0, deletion_comment = False)

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
 
             // Now that we've written to the local array, find out if that array has been
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
```
I see. The race was here. You could add to old table after resize started and someone could have seen it. If the change was not picked up by resize and another writer, wrote to the new table, you would have effectively two values for the same key.

VSadov replied:
Just to check if I understand the problem/solution correctly - it could be a smaller change to add a `bool published` to every table entry and flip that to true after checking for new table. But that would double the size of the table.

---
In reply to: [356353759](https://github.com/dotnet/corert/pull/7906#discussion_r356353759) [](ancestors = 356353759)

VSadov commented on src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:
```diff
@@ -169,38 +185,83 @@ public bool TryGetValue(TKey key, out TValue value)
+            {
+                for (int i = 0; (i < 10000) && value == sentinel; i++)
+                {
+                    value = Volatile.Read(ref hashtable[tableIndex]);
```
>Volatile.Read [](start = 28, length = 13)

This `Volatile.Read` is technically needed -otherwise it is valid to CSE it out of the loop and kill the loop, which would still have correct behavior, in a way, but not what you want.

Other `Volatile.Read`s seem unnecessary.

davidwrighton commented:
@MichalStrehovsky Yes, the pointers table also needs the same sort of fix. I wanted to get some review on the normal table before porting it to the pointers table as its somewhat tricky, and addressing feedback on two implementations isn't super easy.

MichalStrehovsky commented:
Thank you for diagnosing and fixing this David!
